src/main/java/mate/academy/bookshop/validation/IsbnValidator.java

```diff
@@ -0,0 +1,37 @@
+package mate.academy.bookshop.validation;
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import java.util.regex.Pattern;
+import lombok.RequiredArgsConstructor;
+import mate.academy.bookshop.service.BookService;
+import mate.academy.bookshop.util.IsbnFormatter;
+
+@RequiredArgsConstructor
+public class IsbnValidator implements ConstraintValidator<Isbn, String> {
+    private static final String ISBN_VALIDATION_PATTERN
+            = "^(?:ISBN(?:-1[03])?:? )?(?=[0-9X]{10}$|(?=(?:[0-9]+[- ]){3})[- 0-9X]{13}"
+            + "$|97[89][0-9]{10}$|(?=(?:[0-9]+[- ]){4})[- 0-9]{17}$)"
+            + "(?:97[89][- ]?)?[0-9]{1,5}[- ]?[0-9]+[- ]?[0-9]+[- ]?[0-9X]$";
+    private final BookService bookService;
+    private String uniqueMessage;
+
+    @Override
+    public void initialize(Isbn constraintAnnotation) {
+        this.uniqueMessage = constraintAnnotation.uniqueMessage();
+    }
+
+    @Override
+    public boolean isValid(String isbn, ConstraintValidatorContext constraintValidatorContext) {
+        if (isbn == null || !Pattern.compile(ISBN_VALIDATION_PATTERN).matcher(isbn).matches()) {
+            return false;
+        }
+        if (bookService.existByIsbn(IsbnFormatter.format(isbn))) {

```

Comment:
it is a bad practice to use services & repositories in the validators & mappers like that. Usually, they are pretty lightweight and needed for some stuff that can be checked without involving too many resources. So let's move the uniqueness check to the service create/update methods, and after we will do that - it'll be enough to just set a `@Pattern` over the ISBN field in dto or use `@ISBN` as Nazar suggested above